### ISSUE TITLE
Provide TriblerCore builds for ARM Linux

### ISSUE THREAD

**Issue by hbiyik** [Tribler/tribler#5327]
Hello Team

**Problem 1:** A conveinent way for tribler headless (triblerd) should be provided to deploy the server without QT and UI stuff.
Currently run_tribler.py can be run with some ENV variables (CORE_PROCESS etc..) and which is not very conveinent.
Another approach is there run_tirbler.headless.py which is nice way but does not support versioned dirs. I should this part needs to be refactored.

**Problem2:** How to build triblerd for arm variants of Linux. Good news. I already have compiled for ARM! yay, here is how it works:

**ARM buildsys for tribler:**
a debian(wheezy) based image, with arm 32bit soft floating point toolchain (armel triplet) is run under QEMU emulator. All the necessary compenents are built in the emulator (libtorrent, boost, openssl etc.).

This environment gives us a very broad support with >**libc 2.13** and arm support over > armv5, practically **all 32bit arm** socs. (I remember 64bit arms to work with 32bit ISA as well but i need to check).

The downside is, since we are running in an emulator, compilation speed is quite low, but thats ok, we compile once for the dependencies and freezing tribler is not everydays work.

Another downside is, even though soft float arm abi, gives a broad support for SOCs, it runs slower in most of the devices where hard floating point is available (> armv6 basically). This will be fixed with another emulator container with armhf triplet but in first stage i want to work with armel to see the performance and application scope etc..

But to move forward I need a solution for the first problem and decide what to build and deliver.

**Comment by synctext** [comment]
cool progress.
Would it help if we at some point procured a an ARM laptop for our build farm? https://www.pine64.org/pinebook/
Native execution..

**Comment by hbiyik** [comment]
That device is using an arm64, i checked that aarch64 can run isa from armv5 in 32bit as well.

Together with docker containers should help i think, with containers there should be an abstract approach to have both armel, armhf and aarch64 toolchains on the same device.

**Comment by hbiyik** [comment]
I did further investigation and i feel kind of silly for trying to build that qemu image for 3 days since docker has already a feature to emulate the ISA of different CPU architectures and pass to kernel on runtime. This is called binfmt support with QEMU, meaning that you can run guest instructions with full host cpu capabilities in emulation mode without emulating a hardware or SOC. So there is no need for an ARM native device also.

This is something new to me, may be it was there since for a long time but sometimes you learn the hard way :). Nevertheless problem 1 still stands.

**Comment by synctext** [comment]
versioned dirs?
those are the state directories for .Tribler, isolated between upgrades. Feel free to move that code for Triblerd re-usage purposes. Any dev suggestions?

**Comment by hbiyik** [comment]
a side update on buildsys:
https://github.com/hbiyik/triblerd-buildsys/blob/master/dockerfiles/debian-lenny-armel.dockerfile

This container builds the tribler in arm ISA and can work on x86 host, works like a charm.

You can even integrate the foreign CPU python to your IDE (ie eclipse), and run the debugger inside the containe which is awesome, or you can deply the python files of tribler in the container and distribute container. I liked this.

**Comment by hbiyik** [comment]
The buidlsys is ready, and finally i can run tirbler in linux arm, Yay!

With this buildsys, we can run tribler in a linux basically from last 15-20 years!

You can find the buildsys in:

https://github.com/hbiyik/triblerd-buildsys

Also the binaries generated:

https://github.com/hbiyik/triblerd-buildsys/releases/tag/7.5-master

Below is the tollchain information, and version summary used in the compilation progress.

```
{
  "images": {
    "linux-i386": {
      "base": "debian/eol:lenny-slim@sha256:a2230f615ba41081b6727b32234275064e4f53de77dc94765fcaa24489d6f930",
      "openssl-platform": "linux-x86",
      "qemu-platform": "i386",
      "toolchain": {
        "gcc": "4.3.2",
        "glibc": "2.7"
      }
    },
    "linux-amd64": {
      "base": "debian/eol:lenny-slim@sha256:4211c70bd47b74942957b208e9f54cb52eaa15b0b48c3d76f3f9b502929d1d01",
      "openssl-platform": "linux-x86_64",
      "qemu-platform": "x86_64",
      "toolchain": {
        "gcc": "4.3.2",
        "glibc": "2.7"
      }
    },
    "linux-armel": {
      "base": "debian/eol:lenny-slim@sha256:859461d1c5e8e39244a803fef8cad2b840cd9198542d7edc0bada781f5794b1a",
      "openssl-platform": "linux-generic32",
      "qemu-platform": "arm",
      "toolchain": {
        "gcc": "4.3.2",
        "glibc": "2.7"
      }
    },
    "linux-armhf": {
      "base": "debian/eol:wheezy-slim@sha256:54a2fa7015ca274803d8e64f622a375057efa89d3595bb9148d62aa657b3b3c8",
      "openssl-platform": "linux-armv4",
      "qemu-platform": "arm",
      "toolchain": {
        "gcc": "4.6.3",
        "glibc": "2.13"
      }
    },
    "linux-aarch64": {
      "base": "debian/eol:jessie-slim@sha256:600ca584836c903364017aea8f4ca4af335f04c949a7dd78801477cf3cb6cdc5",
      "openssl-platform": "linux-aarch64",
      "qemu-platform": "aarch64",
      "toolchain": {
        "gcc": "4.9.2",
        "glibc": "2.19"
      }
    }
  },
  "deps": {
    "perl": {
      "version": "5.30.2",
      "dir": "sources/perl-5.30.2"
    },
    "openssl": {
      "version": "1.1.1",
      "dir": "sources/openssl-1.1.1g"
    },
    "Python": {
      "version": "3.7.7",
      "dir": "sources/Python-3.7.7"
    },
    "boost": {
      "version": "1_53_0",
      "dir": "sources/boost_1_53_0"
    },
    "libtorrent_rasterbar": {
      "version": "1.1.5",
      "dir": "sources/libtorrent-rasterbar-1.1.5"
    },
    "libsodium": {
      "version": "1.0.18",
      "dir": "sources/libsodium-1.0.18"
    },
    "cryptography": {
      "version": "2.9.2",
      "dir": "sources/cryptography-2.9.2"
    },
    "sqlite_autoconf": {
      "version": "3320100",
      "dir": "sources/sqlite-autoconf-3320100"
    }
  },
  "interpreter": "python3.7"
}
```

**Comment by hbiyik** [comment]
@devos50 

How is the current buildsystem for windows binaries? Is there any docker image, i think a linux container over wine + python3.7 is possible..

**Comment by devos50** [comment]
@hbiyik we are using [PyInstaller](https://www.pyinstaller.org) to bundle the Python executable/runtime, and create an installer with NSIS.

**Comment by hbiyik** [comment]
I mean is there any compilation environment like docker like there is for linux? Or precompiled binaries are used for openssl, boost, libtorrent, libsodium, or some windows machine?

**Comment by devos50** [comment]
No, we use pre-compiled libraries. We have [compilation instructions](https://tribler.readthedocs.io/en/devel/development/development.html#windows), though.

**Comment by hbiyik** [comment]
Ok, i see, I have a funny experiment to compile all deps in a linux container, and freezing with wine, so that each platform release will have the same c code, lets see how that will work

**Comment by devos50** [comment]
We can test the outcome of your experiment on our Windows machines if you wish :)

**Comment by hbiyik** [comment]
Would be great, I have also made some experiments on android as well, successfully builds and it kind of works, but i expect bunch of errors on runtime when all the test suite is run mainly from psutil and some other libs.

![image](https://user-images.githubusercontent.com/24766436/83851849-3a25e280-a713-11ea-9fe3-c83aff8157f7.png)

**Comment by hbiyik** [comment]
Very good news. I have amanged to run unittests in android as well. 

6 of 659 tests failed. Mainly due to psutil, and some netowrk socket related things, should be so hard to wrap around it.

I think android port is almost ready, there is not much of workload here.

Here is the overall output and traceback of all tests:

**Tribler Core: 659 Tests, 6 Failed
Anydex : 267 Tests. 12 Failed (due to bitcoinlib, those tests dont work on pc also so no issues on android)
IPv8: There are some import isseues there, could not run yet, but dont expect a biggie
Total tested: 926**
https://pastebin.com/raw/5cErn8DJ

```

**Comment by hbiyik** [comment]
@synctext @ichorid @devos50 @egbertbouman 
I think i am finished with the buildsystem, and can finally produce binaries for arm/x86 linux and arm/x86 android.

Could you take your time and read the readme file in 
https://github.com/hbiyik/triblerd-buildsys

which details why, how and what.

- **linux-i386:** practically for any linux distro released more than 10 years ago that that runs on a x64 cpu etiher in 32bit or 64 bit mode
- **linux-amd64:** same as above but only in 64 bit mode, with better performance and assembly level optimizations
- **linux-armel:** very old arm based linux distros down to >= armv5
- **linux-armhf:** most arm7 based devices and linux devices regardless of what kind of SOC it is.
- **linux-aarch64:** nearly all distros with armv8 SOCs.
- **android-i386:** atom based devices with Android >= 5.0 Lollipop
- **android-amd64:** atom based devices with Android >= 5.0 Lollipop in 64bit mode
- **android-armhf:** all android devices with arm>=v7 and Android >=5.0 Lollipop 
- **android-aarch64:** all android devices with arm>=v8 and Android >=5.0 Lollipop

https://github.com/hbiyik/tribler/commit/1d4ed4ad260ee78cd40c39c3e0bcd838b338d3d9

**Comment by synctext** [comment]
wow. Very elaborate documentation. Worked with cross compilers, but this is not easy to grasp. How difficult is a download/upload performance table to make for a few boards?

**Comment by hbiyik** [comment]
I tried to put a lot of detail in the documentation cos i know it is quite a tricky topic and some of the things are based on historical event chains (like glibc dependencies), in case a student may take over, he can have a holy bible to consult instead of praying :)

but never the less all this complexity is reduced to some basic commands in the case of usage.

I tested with an armv8 cpu and the speed can easily get over 1MByte/s with 1 hop encrypted tunnels. What would be great is to have a unit testing for those benchmarks so the results would be comperable.

I have a pi-1, which is great to test the behaviour in absolute lowest device, 2 android phones with decent 8core cpus.

**Comment by synctext** [comment]
The complexity is quite something, your expertise is vital for maintenance. Kodi is a big new community, millions of users. Some would ask us for their favourite new feature.

1 MByte/sec with 1 tunnel hop, impressive results. We have numerous loose Android devices (10+), but no testing infrastructure. Unclear to me how far away you are from .APK builds. Like for high-end octa-core (4x2.4 GHz Cortex-A73 & 4x1.8 GHz Cortex-A53).

**Comment by hbiyik** [comment]
I can always support in such cases, no issues.

Anyone with a linux x64 pc can already build the apk with this already. For testing, before bundling the core into apk, better to patch this line from localhost to 0.0.0.0 

https://github.com/Tribler/tribler/blob/51dc33b09846fdd533df3ccd2fa7c25b397629bb/src/tribler-core/tribler_core/restapi/rest_manager.py#L100

and after running the apk, control the core over ip.of.the.device:8085 from a desktop pc and do fancy testing.

See below how core is initialized in android, it is basically triblerr_headless.py

https://github.com/hbiyik/triblerd-buildsys/blob/master/buildozer/main.py

**Comment by ichorid** [comment]
Now the question is: how do we integrate this into our devops/release cycle?...

**Comment by devos50** [comment]
@hbiyik impressive work! Sorry for the delayed response; we have been/are quite busy with scientific output the past few weeks so engineering efforts were a bit low. This line of work therefore did not get the attention that it deserves. From experience, I know that working with P4A/cross-compilation is not easy and we always struggled to get a proper build process for Android up and running. I will check out the readme and try to get it up and running later today.

Integration in our Jenkins pipelines should be the next step, and most likely requires further discussion/granting you some permissions to our infrastructure. We used to have a few Android phones hooked up to one of our computers, but they died recently. Can we run the unit tests in an Android simulator first? After some infrastructure improvements, I hope we can run these tests on a real device. 👍 

If you wish, I can invite you to our Slack channel so we can have a more real-time discussion about this (please send me an email if you want this invitation).

Also, note that one of our top master students is working on a [Kotlin port of Tribler](https://github.com/Tribler/trustchain-superapp) in parallel. You might want to check out that app too.

**Comment by devos50** [comment]
@hbiyik Also, I suggest moving the repository to the Tribler organization, if you are ok with that 👍

**Comment by hbiyik** [comment]
@devos50 
Ok i think i need access rights for this transfer thing, or you can fork it if you want.

>  I know that working with P4A/cross-compilation is not easy and we always struggled to get a proper build process for Android up and running

I know what you mean i saw a lot of foot prints of @brussee in p4a repo, thanks to him he has already done the heavy lifting. But things have changed and now we have an excellent tool like docker, where we can freeze the environment and move it around freely, and even emulate remote foreign cpu instructuion sets. Now things are much way easier.

> Can we run the unit tests in an Android simulator first?

If theere is adb access then you should yet i tested on real devices and it definetely works, after some fixing obvious stuff in tribler source. (mentioned in above comments)

>> Can we run the unit tests in an Android simulator first?

Would love to, add me at huseyinbiyik at  hotmail.

@ichorid 
Firstly I would like to point out  another use case for this buildsystem is, how tribler application is distributed.

The generated core binaries are single files, with a filesize around 25 ~ 35 MB.

It is possible to distribute the QT gui only, and the gui itself may check the binary repo and download the latest core binary and execute it as a service. This way whenever there is a new version of core, the user wont need to update the Tribler but tribler will automatically do it, and this will significantly reduce the release cycle time, and help to overcome user resistance on system upgrades.

**Comment by ichorid** [comment]
>It is possible to distribute the QT gui only, and the gui itself may check the binary repo and download the latest core binary and execute it as a service. This way whenever there is a new version of core, the user wont need to update the Tribler but tribler will automatically do it, and this will significantly reduce the release cycle time, and help to overcome user resistance on system upgrades.

That is a very interesting idea, indeed. 

Currently, there is a variety of methods of application deployment on Linux platforms (e.g. Snap vs Docker vs package manager vs pip vs whatever...) We should carefully select the most efficient one or two and support them. The problem with this stuff is that supporting a variety of methods will spread the team's effort even thinner than it is now, and can confuse users.

We would prefer not to touch the existing build system until fresh blood arrives in September, though.

**Comment by hbiyik** [comment]
My motivation was to have a uniform way of having a deployment mechanism of the daemon for kodi users, and came up with this, I will surely use this method because it really helps to reduce all hose complexity in case of a new tribler release, may be i can use this as a trial in kodi user base and collect the feedback. Since http serving over github is totally free :)

**Comment by hbiyik** [comment]
just added windows to toolchains.

now 2 more targets are available

**windows-i386
windows-amd64**

only thing that this buildsystem cant package to from the same linux distro is now OSX. I would like play dead when apple stuff is in place :)

**Comment by qstokkink** [comment]
With a `musl` build and an `aarch64` build available, I think this is where we want to leave it for "officially published and supported" multi-arch. We simply don't have the team size to support more.
